include/zio/zio_defs.h

```diff
@@ -0,0 +1,510 @@
+ * | | | | Channel Attribute 2----------------------------------------+ || | |
+ * | | | | |                   CHAN_ATTR_SAMP_FREQ                    | || | |
+ * | | | | +----------------------------------------------------------+ || | |
+ * | | | +--------------------------------------------------------------+| | |
+ * | | +-----------------------------------------------------------------+ | |
+ * | | |Channel 1                                                        | | |
+ * | | +-----------------------------------------------------------------+ | |
+ * | | |Channel 2                                                        | | |
+ * | | +-----------------------------------------------------------------+ | |
+ * | +---------------------------------------------------------------------+ |
+ * +-------------------------------------------------------------------------+
+ */
+
+/**
+ * @brief ZIO device channel types.
+ *
+ * Each channel type is associated with a specific SI unit type, described
+ * in the inline comments for each enum entry.
+ */
+typedef enum zio_dev_chan_type {
+	/**
+	 * Raw data channel.
+	 *
+	 * No SI type is prescribed for this channel.
+	 */
+	DEV_CHAN_RAW,
+
+	/**
+	 * Voltage.
+	 *
+	 * Expressed as volts (V).
+	 */
+	DEV_CHAN_VOLTAGE,
+
+	/**
+	 * Current.
+	 *
+	 * Expressed as milli-amps (mA).
+	 */
+	DEV_CHAN_CURRENT,
+
+	/**
+	 * Resistance.
+	 *
+	 * Expressed as ohms.
+	 */
+	DEV_CHAN_RESISTANCE,
+
+	/**
+	 * Capacitance.
+	 *
+	 * Expressed as picofarad (pF).
+	 */
+	DEV_CHAN_CAPACITANCE,
+
+	/**
+	 * Inductance.
+	 *
+	 * Expressed as microhenries (uH).
+	 */
+	DEV_CHAN_INDUCTANCE,
+
+	/**
+	 * Impedance.
+	 *
+	 * Expressed as ohms (Z).
+	 */
+	DEV_CHAN_IMPEDANCE,
+
+	/**
+	 * Frequency.
+	 *
+	 * Expressed as Hertz (Hz).
+	 */
+	DEV_CHAN_FREQUENCY,
+
+	/**
+	 * Amplitude.
+	 *
+	 * Note: While amplitude is a unitless value with a floating-point
+	 * range of -1.0 .. 1.0, you can associate this with a range of integer
+	 * values by means of the 'CHAN_ATTR_RAW_DATA' attribute. Normalised
+	 * floating-point values should be assigned to 'CHAN_ATTR_SI_DATA',
+	 * but the source integer values, such as 's16_t', can be provided
+	 * as raw data to avoid expensive floating-point operations if
+	 * desired.
+	 *
+	 * Expressed as a unitless value in the range of -1.0 to 1.0.
+	 */
+	DEV_CHAN_AMPLITUDE,
+
+	/**
+	 * Generic temperature.
+	 *
+	 * Expressed as degrees celsius (C).
+	 */
+	DEV_CHAN_TEMP,
+
+	/**
+	 * Ambient temperature.
+	 *
+	 * Expressed as degrees celsius (C).
+	 */
+	DEV_CHAN_TEMP_AMBIENT,
+
+	/**
+	 * Die temperature.
+	 *
+	 * Expressed as degrees celsius (C).
+	 */
+	DEV_CHAN_TEMP_DIE,
+
+	/**
+	 * Single-axis acceleration.
+	 *
+	 * Expressed as meters per second squared (m/s^2).
+	 */
+	DEV_CHAN_ACCEL,
+
+	/**
+	 * Three-axis (XYZ) acceleration.
+	 *
+	 * Expressed as meters per second squared (m/s^2).
+	 */
+	DEV_CHAN_ACCEL_VECTOR,
+
```

Comment:
I think that the number of axis should be specified in a different parameter.
The channel type is always ACCEL (or GYRO or MAGN)

Reply:
It does and I'm not in love with the concept of vector, my concern is that with sensor fusion you often want to be sure that the three values from from the same sample or conversion process, and if you have three different channels for X+Y+Z, you need to add a means to sync those values with each other, which becomes more code to validate. The need to group multiple channels will come up elsewhere, of course, such as in light sensors with multiple photo diodes, each with their own channel, though that takes the approach of having a single channel per photodiode and the attributes fill the details in.

How would you see the problem of needing to keep X/Y/Z related to each other so that you don't get X+Y from one conversion, but Z from the next?

Reply:
> How would you see the problem of needing to keep X/Y/Z related to each other so that you don't get X+Y from one conversion, but Z from the next?

I need to see more carefully for theample the Google CHRE, where they define the sensor type
as ACCEL or GYRO or MAGN only, but also define in the sensor property the number of axis
supported by the sensor (One of SINGLE, TRIPLE, EMBEDDED, with latter one used for
temp/baro/humidity/stepc and so on).

Then a sensor when generates samples enqueue above them as events, and a general task
would consume them. This is different from how zephyr works now where samples are pulled
from above and not pushed from below as events.

Reply:
If you have some suggestions here, it will never be easier than now to make the right decisions and tradeoffs!

Reply:
Yes, sure.
I'll try to get back to this asap.